dali/python/nvidia/dali/shared_batch.py

```diff
@@ -0,0 +1,262 @@
+                   writer.meta_data_size)
+
+
+def deserialize_sample(buffer: shared_mem.SharedMem, sample):
+    if isinstance(sample, SampleMeta):
+        offset = sample.offset
+        assert offset % sample.dtype.itemsize == 0, "Sample offset is misaligned."
+        buffer = buffer.buf[offset:offset + sample.nbytes]
+        return np.ndarray(sample.shape, dtype=sample.dtype, buffer=buffer)
+    if isinstance(sample, (tuple, list,)):
+        return type(sample)(deserialize_sample(buffer, part) for part in sample)
+    return sample
+
+
+def deserialize_sample_meta(buffer: shared_mem.SharedMem, shared_batch_meta: SharedBatchMeta):
+    """Helper to deserialize SampleMeta from memory based on SharedBatchMeta.
+    """
+    sbm = shared_batch_meta
+    if sbm.meta_size == 0:
+        return []
+    pickled_meta = buffer.buf[sbm.meta_offset:sbm.meta_offset + sbm.meta_size]
+    samples_meta = pickle.loads(pickled_meta)
+    return samples_meta
+
+
+def deserialize_batch(buffer: shared_mem.SharedMem, shared_batch_meta: SharedBatchMeta):
+    """Deserialize samples from the smem buffer and SampleMeta descriptions.
+
+    Parameters
+    ----------
+    buffer : shared_mem.SharedMem
+        Buffer with serialized sample data
+    shared_batch_meta : SharedBatchMeta
+        Metadata about serialized data in memory
+
+    Returns
+    -------
+    List of (idx, numpy array) or (idx, tuple of numpy arrays)
+        List of indexed deserialized samples
+    """
+    samples = deserialize_sample_meta(buffer, shared_batch_meta)
+    return [(idx, deserialize_sample(buffer, sample)) for (idx, sample) in samples]
+
+
+class SharedMemChunk:
+    """Simple wrapper around shared memory chunks. Most importantly adds mem_chunk_id used
+    to identify chunks in the communication between parent and worker process
+    (shared memory handles/file descriptors cannot serve this purpose easily as the same
+    mapped memory chunk can have different handles in both processes).
+    """
+
+    def __init__(self, mem_chunk_id: str, capacity: int):
+        # mem_chunk_id must be unique among all workers and callbacks in the pool,
+        # used to identify shared memory chunks in the communication between processes
+        self.mem_chunk_id = mem_chunk_id
+        self.shm_chunk = shared_mem.SharedMem.allocate(capacity)
+        self.capacity = capacity
+
+    def resize(self, new_capacity):
+        self.shm_chunk.resize(new_capacity, trunc=True)
+        self.capacity = new_capacity
+
+    def close(self):
+        # Shared memory handle and mapping will be freed automatically when shm pybind wrapper gets
+        # garbage collected anyway, but you can be so nice to free them as soon as you know
+        # you won't use them
+        self.shm_chunk.close()
+
+
+def _to_numpy(sample):
+    if isinstance(sample, np.ndarray):
+        return sample
+    if types._is_mxnet_array(sample):
+        return sample.asnumpy()
```

Comment:
Won't this convert data to CPU? Maybe you should check if it is a GPU tensor similarly to Torch case?

Reply:
Probably it would, added similar check.